### PR TITLE
PHDO-122 - Updates to graphql report mutation

### PR DESCRIPTION
**Pull request by mkrystof**
- Utilizes the schema validation library for validating reports.
- Provides a result in the graphql response when a report is ingested via a graphql mutation, supporting both adding and replacing.

An example mutation for replacing a report is below.

```graphql
mutation {
    upsertReport(
        action: "replace"
        report: {
            report_schema_version: "1.0.0"
            upload_id: "e2b3a950-4c2c-4756-868e-f81ec27f27fb"
            user_id: "test-user2"
            data_stream_id: "dex-testing"
            data_stream_route: "test-event1"
            jurisdiction: "SMOKE5"
            sender_id: "APHL"
            dataProducer_id: "smoke-test-data-producer"
            dex_ingest_datetime: "2024-07-10T15:40:10.162+00:00"
            status: "SUCCESS"
            message_metadata: {
                message_uuid: "5a1fff57-2ea1-4a64-81de-aa7f3096a1ce"
                message_hash: "38c2cc0dcc05f2b68c4287040cfcf71"
                aggregation: "SINGLE"
                message_index: 1
            }
            stage_info: {
                service: "UPLOAD API"
                action: "upload-completed"
                version: "0.0.49-SNAPSHOT"
                status: "SUCCESS"
                issues: null
                start_processing_time: "2024-07-10T15:40:10.162+00:00"
                end_processing_time: "2024-07-10T15:40:10.228+00:00"
            }
            tags: { tag_field1: "value1" }
            data: { data_field1: "value1" }
            content_type: "application/json"
            content: {
                content_schema_name: "upload-completed"
                content_schema_version: "1.0.0"
                status: "SUCCESS"
            }
        }
    ) {
        issues
        reason
        reportId
        result
        schemaFileNames
        uploadId
    }
}
```

### CODE REVIEW

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt [hunk 1]
```diff
@@ -63,277 +88,116 @@ class ReportMutationService : KoinComponent {
...
-                data = input.data?.associate { it.key to it.value },
-                contentType = input.contentType,
-                jurisdiction = input.jurisdiction,
-                senderId = input.senderId,
-                dataProducerId = input.dataProducerId,
-                content = parsedContent,
-                timestamp = input.timestamp
+    private fun validateReport(input: Map<String, Any?>?): ValidatedReportResult {
+        try {
+            logger.info("The report received: $input and will be converted to Json")
```
Comment (manu-govind) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt:142]: Should we check whether input is null, here?

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt [hunk 2]
```diff
@@ -1,29 +1,56 @@
...
+import gov.cdc.ocio.reportschemavalidator.service.SchemaValidationService
+import gov.cdc.ocio.reportschemavalidator.utils.DefaultJsonUtils
+import gov.cdc.ocio.reportschemavalidator.validators.JsonSchemaValidator
+import mu.KLogger
 import mu.KotlinLogging
 import org.koin.core.component.KoinComponent
 import org.koin.core.component.inject
+import java.time.Instant
+import java.time.OffsetDateTime
 import java.util.*
```
Comment (manu-govind) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt:34]: Header summary missing.

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt [hunk 3]
```diff
@@ -63,277 +88,116 @@ class ReportMutationService : KoinComponent {
      * @throws ContentException If there is an error with the content format.
      */
     @Throws(BadRequestException::class, ContentException::class, Exception::class)
```
Comment (manu-govind) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt:90]: Function header missing

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt [hunk 4]
```diff
@@ -63,277 +88,116 @@ class ReportMutationService : KoinComponent {
...
-                stageInfo = input.stageInfo?.let { mapInputToStageInfo(it) },
-                tags = input.tags?.associate { it.key to it.value },
-                data = input.data?.associate { it.key to it.value },
-                contentType = input.contentType,
-                jurisdiction = input.jurisdiction,
-                senderId = input.senderId,
-                dataProducerId = input.dataProducerId,
-                content = parsedContent,
-                timestamp = input.timestamp
+    private fun validateReport(input: Map<String, Any?>?): ValidatedReportResult {
```
Comment (manu-govind) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt:140]: Function header missing

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/CustomSchemaGeneratorHooks.kt [hunk 5]
```diff
@@ -1,15 +1,135 @@
...
+/**
+ * Converts a [Value] generic to a [JsonNode].
+ *
+ * @param value Value<*>
+ * @return JsonNode
+ */
+private fun convertValue(value: Value<*>): JsonNode {
+    return when (value) {
+        is StringValue -> JsonNodeFactory.instance.textNode(value.value)
+        is IntValue -> JsonNodeFactory.instance.numberNode(value.value)
```
Comment (manu-govind) [inline pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/CustomSchemaGeneratorHooks.kt:87]: Should we check for "double" type as well?

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt [hunk 6]
```diff
@@ -349,27 +213,25 @@ class ReportMutationService : KoinComponent {
...
-        if (report.id != null) {
-            throw BadRequestException("ID should not be provided for create action.")
-        }
-
-        // Validate uploadId
-        if (report.uploadId.isNullOrBlank()) {
-            throw BadRequestException("Upload ID must be provided.")
-        }
+    @Throws(BadRequestException::class, ContentException::class)
+    private fun createReport(report: Report) { // TODO - move to ReportManager.kt
```
[the remark this answers is not in the document]
Comment (mkrystof) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt]: Done

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt (hunk 2 above)
Comment (mkrystof) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt:31]: Done

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/utils/ReportManager.kt
[the remark this answers is not in the document]
Comment (mkrystof) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/utils/ReportManager.kt:1]: Yes, I kept it and moved the report management functions into it (create, replace report).

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt (hunk 3 above)
Comment (mkrystof) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt:88]: Done

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt (hunk 4 above)
Comment (mkrystof) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt:144]: Done

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/CustomSchemaGeneratorHooks.kt (hunk 5 above)
Comment (mkrystof) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/plugins/CustomSchemaGeneratorHooks.kt:87]: We don't need a "DoubleValue" one here, which doesn't exist.  The FloatValue one under the hood is a BigDecimal type, which covers doubles and floats.

File: pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt (hunk 1 above)
Comment (mkrystof) [reply pstatus-graphql-ktor/src/main/kotlin/gov/cdc/ocio/processingstatusapi/services/ReportMutationService.kt:151]: Added null or empty check on the report input.

**Review by manu-govind — APPROVED** [review]
Approved

**Pull request merged**